### PR TITLE
Fix coverity errors in heap parsing code and improve heap API

### PR DESCRIPTION
**Pull request by MalhotraPulak**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

- Fixing coverity errors of previous PR
- Improve the Heap API for Cutter heap widget

I will be updating the Heap API as I work on the Cutter Heap widgets and need new API calls/update the old ones
My commits to cutter can be found [here](https://github.com/rizinorg/cutter/commits/heap-dock-widget)
...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

**Comment by XVilka** [comment]
See also this warning please:
```
[1161/1660] Compiling C object librz/core/librz_core.so.0.3.0-git.p/task.c.o
In file included from ../librz/core/task.c:5:
In file included from ../librz/include/rz_core.h:34:
../librz/include/rz_heap_glibc.h:34:9: warning: 'MAX' macro redefined [-Wmacro-redefined]
#define MAX(a, b)              (((a) > (b)) ? (a) : (b))
        ^
/usr/include/sys/param.h:315:9: note: previous definition is here
#define MAX(a,b) (((a)>(b))?(a):(b))
        ^
1 warning generated.
```
Just use `RZ_MAX`

File: librz/core/linux_heap_glibc.c [hunk 1]
```diff
@@ -1249,6 +1252,12 @@ void GH(print_malloc_states)(RzCore *core, GHT m_arena, MallocState *main_arena,
...
 			ut64 ta_addr = ta->GH(next);
 			if (!GH(rz_heap_update_main_arena)(core, ta->GH(next), ta)) {
+				if (json) {
+					pj_end(pj);
+					pj_end(pj);
+					rz_cons_println(pj_string(pj));
+					pj_free(pj);
+				}
 				free(ta);
 				return;
```
Comment (ret2libc) [inline librz/core/linux_heap_glibc.c]: maybe `goto end` and put `end:` near the end of the function where all these things are already done?

File: librz/core/linux_heap_glibc.c [hunk 2]
```diff
@@ -2013,7 +2021,13 @@ RZ_IPI RzCmdStatus GH(rz_cmd_heap_chunks_print_handler)(RzCore *core, int argc,
 			rz_cons_printf("f %s %d 0x%" PFMT64x "\n", name, (int)pos->size, (ut64)pos->addr);
 			free(name);
 		} else if (mode == RZ_OUTPUT_MODE_LONG_JSON) { // graph
+			if (node_title) {
+				free(node_title);
+			}
```
Comment (ret2libc) [inline librz/core/linux_heap_glibc.c]:
```suggestion
			free(node_title);
```

File: librz/core/linux_heap_glibc.c [hunk 3]
```diff
@@ -2013,7 +2021,13 @@ RZ_IPI RzCmdStatus GH(rz_cmd_heap_chunks_print_handler)(RzCore *core, int argc,
...
 			node_title = rz_str_newf("  Malloc chunk @ 0x%" PFMT64x " ", (ut64)pos->addr);
+			if (node_data) {
+				free(node_data);
+			}
```
Comment (ret2libc) [inline librz/core/linux_heap_glibc.c]:
```suggestion
			free(node_data);
```

File: librz/core/linux_heap_glibc.c [hunk 4]
```diff
@@ -1959,13 +1961,18 @@ RZ_IPI RzCmdStatus GH(rz_cmd_heap_chunks_print_handler)(RzCore *core, int argc,
...
 	}
 	RzANode *top = RZ_EMPTY, *chunk_node = RZ_EMPTY, *prev_node = RZ_EMPTY;
-	char *top_title, *top_data, *node_title, *node_data;
+	char *top_title = NULL, *top_data = NULL, *node_title = NULL, *node_data = NULL;
 	bool first_node = true;
 	top_data = rz_str_new("");
-	top_title = rz_str_new("");
 	if (mode == RZ_OUTPUT_MODE_JSON) {
 		if (!pj) {
+			rz_cons_newline();
```
Comment (ret2libc) [inline librz/core/linux_heap_glibc.c]: goto?

**Review by XVilka — APPROVED** [review]

File: librz/core/linux_heap_glibc.c [hunk 5]
```diff
@@ -1233,6 +1233,7 @@ void GH(print_malloc_states)(RzCore *core, GHT m_arena, MallocState *main_arena,
 		PRINTF_YA("0x%" PFMT64x, (ut64)main_arena->GH(next));
 		rz_cons_printf(")\n");
 	} else {
+		pj = pj_new();
```
Comment (wargio) [inline librz/core/linux_heap_glibc.c:1240]: check if pj is not NULL

File: librz/core/linux_heap_glibc.c [hunk 6]
```diff
@@ -1617,42 +1623,57 @@ static void GH(print_main_arena_bins)(RzCore *core, GHT m_arena, MallocState *ma
...
 	}
 	// main arena
-	GH(rz_heap_update_main_arena)
-	(core, m_arena, ta);
-	rz_list_append(arena_list, ta);
+	if (!GH(rz_heap_update_main_arena)(core, m_arena, ta)) {
+		free(ta);
+		return arena_list;
+	}
+	RzArenaListItem *item = RZ_NEW0(RzArenaListItem);
```
Comment (wargio) [inline librz/core/linux_heap_glibc.c:1649]: check if item is not null

File: librz/core/linux_heap_glibc.c [hunk 7]
```diff
@@ -1617,42 +1623,57 @@ static void GH(print_main_arena_bins)(RzCore *core, GHT m_arena, MallocState *ma
...
 		while (GH(is_arena)(core, m_arena, ta->GH(next)) && ta->GH(next) != m_arena) {
 			ut64 ta_addr = ta->GH(next);
 			ta = RZ_NEW0(MallocState);
 			if (!GH(rz_heap_update_main_arena)(core, ta_addr, ta)) {
+				free(ta);
 				return arena_list;
 			}
 			// thread arenas
-			rz_list_append(arena_list, ta);
+			item = RZ_NEW0(RzArenaListItem);
```
Comment (wargio) [inline librz/core/linux_heap_glibc.c:1668]: same here.

File: librz/core/linux_heap_glibc.c [hunk 8]
```diff
@@ -1821,27 +1834,33 @@ RZ_API RzList *GH(rz_heap_chunks_list)(RzCore *core, MallocState *main_arena,
...
 		}
-
 		RzHeapChunkListItem *block = malloc(sizeof(RzHeapChunkListItem));
 		block->addr = prev_chunk_addr;
 		block->status = status;
 		block->size = prev_chunk_size;
 		rz_list_append(chunks, block);
 	}
+	if (top_chunk) {
+		RzHeapChunkListItem *block = malloc(sizeof(RzHeapChunkListItem));
```
Comment (wargio) [inline librz/core/linux_heap_glibc.c]: use `RZ_NEW(RzHeapChunkListItem)` and check if block is not null.

**Review by wargio — CHANGES_REQUESTED** [review]

**Review by XVilka — CHANGES_REQUESTED** [review]
Could you please also add doxygen documentation for all newly added `RZ_API` functions?

File: librz/include/rz_core.h [hunk 9]
```diff
@@ -774,21 +774,27 @@ RZ_API void rz_core_sysenv_end(RzCore *core, const char *cmd);
...
-RZ_API bool rz_heap_update_main_arena_64(RzCore *core, ut64 m_arena, MallocState *main_arena);
-RZ_API RzList *rz_heap_tcache_list_64(RzCore *core, ut64 m_arena, MallocState *main_arena, bool main_thread_only);
-RZ_API RzHeapChunk_32 *rz_heap_get_chunk_at_addr_32(RzCore *core, ut32 addr);
-RZ_API RzList *rz_heap_bin_content_list_32(RzCore *core, MallocState *main_arena, int bin_num);
-RZ_API RzList *rz_heap_arenas_list_32(RzCore *core, ut32 m_arena, MallocState *main_arena);
-RZ_API RzList *rz_heap_heap_chunks_list_32(RzCore *core, MallocState *main_arena, ut32 m_arena, ut32 m_state);
-RZ_API bool rz_heap_resolve_main_arena_32(RzCore *core, ut32 *m_arena);
-RZ_API bool rz_heap_update_main_arena_32(RzCore *core, ut32 m_arena, MallocState *main_arena);
-RZ_API RzList *rz_heap_tcache_list_32(RzCore *core, ut32 m_arena, MallocState *main_arena, bool main_thread_only);
+/* cmd_linux_heap_glibc.c */
```
Comment (XVilka) [inline librz/include/rz_core.h:777]: Please move doxygen to implementation (C file) place instead of the header.

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/core/linux_heap_glibc.c [hunk 10]
```diff
@@ -2303,4 +2347,78 @@ RZ_IPI RzCmdStatus GH(rz_cmd_heap_arena_bins_print_handler)(RzCore *core, int ar
...
+	if (!main_arena) {
+		return rz_list_newf(free);
+	}
+	if (!GH(rz_heap_update_main_arena)(core, m_state, main_arena)) {
+		free(main_arena);
+		return rz_list_newf(free);
+	}
+	return GH(rz_heap_chunks_list)(core, main_arena, m_arena, m_state, true);
+}
+/**
```
Comment (XVilka) [inline librz/core/linux_heap_glibc.c:2399]: Missing newline.

File: librz/core/linux_heap_glibc.c [hunk 11]
```diff
@@ -2303,4 +2347,78 @@ RZ_IPI RzCmdStatus GH(rz_cmd_heap_arena_bins_print_handler)(RzCore *core, int ar
...
+	if (!main_arena) {
+		return rz_list_newf(free);
+	}
+	if (!GH(rz_heap_update_main_arena)(core, m_arena, main_arena)) {
+		free(main_arena);
+		return rz_list_newf(free);
+	}
+	return GH(rz_heap_arenas_list)(core, m_arena, main_arena);
+}
+/**
```
Comment (XVilka) [inline librz/core/linux_heap_glibc.c:2374]: Missing newline.

File: librz/core/linux_heap_glibc.c [hunk 12]
```diff
@@ -2303,4 +2347,78 @@ RZ_IPI RzCmdStatus GH(rz_cmd_heap_arena_bins_print_handler)(RzCore *core, int ar
 	(core, m_state, main_arena, global_max_fast, bin_format, json);
 	free(main_arena);
 	return RZ_CMD_STATUS_OK;
+}
```
Comment (XVilka) [inline librz/core/linux_heap_glibc.c:2351]: Missing newline

**Review by XVilka — APPROVED** [review]
Please address missing newlines between functions and lets merge this.

File: librz/core/cmd_linux_heap_glibc.c [hunk 13]
```diff
@@ -50,4 +50,32 @@ RZ_IPI int rz_cmd_heap_fastbins_print(void *data, const char *input) {
 
 RZ_IPI RzCmdStatus rz_cmd_heap_arena_bins_print_handler(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
 	call_handler(rz_cmd_heap_arena_bins_print_handler, argc, argv, mode);
+}
+
+/* API functions for Heap Viewer in Cutter */
```
Comment (ret2libc) [inline librz/core/cmd_linux_heap_glibc.c:55]: why are they in `cmd_linux_heap_glibc` if they are not commands?

File: librz/core/linux_heap_glibc.c [hunk 14]
```diff
@@ -1132,6 +1135,7 @@ RZ_API RzList *GH(rz_heap_tcache_list)(RzCore *core, GHT m_arena, MallocState *m
 		return tcache_list;
 	}
 	if (!GH(tcache_read)(core, tcache_start, rz_tcache)) {
+		free(rz_tcache);
```
Comment (ret2libc) [inline librz/core/linux_heap_glibc.c]: maybe it's `GH(tcache_free)`?

**Review by wargio — APPROVED** [review]
besides ret2libc comments looks good

File: librz/core/cmd_linux_heap_glibc.c (hunk 13 above)
Comment (MalhotraPulak) [reply librz/core/cmd_linux_heap_glibc.c:55]: I cannot put them in `linux_heap_glibc.c` as they are not supposed to wrapped in the `GH` macro which is used across that file. Also, `cmd_linux_heap_glibc` uses the `call_handler` macro which I needed for these API  calls too. Do you have any other suggestions where I should put these?

**Comment by yossizap** [comment]
>I will be updating the Heap API as I work on the Cutter Heap widgets and need new API calls/update the old ones

Is there anything still missing for Cutter or can this move out of WIP?

**Comment by MalhotraPulak** [comment]
> > I will be updating the Heap API as I work on the Cutter Heap widgets and need new API calls/update the old ones
> 
> Is there anything still missing for Cutter or can this move out of WIP?

@yossizap There are still a few API calls I need to modify/add. Mostly for adding support for bins and tcache to the heap widget in Cutter.
I can make another PR for them if we want to merge this now

**Comment by yossizap** [comment]
> I can make another PR for them if we want to merge this now

Since this PR was already reviewed and existing rizin functionality is intact after these changes I think that's the best option.

File: librz/core/linux_heap_glibc.c [hunk 15]
```diff
@@ -1132,6 +1135,8 @@ RZ_API RzList *GH(rz_heap_tcache_list)(RzCore *core, GHT m_arena, MallocState *m
 		return tcache_list;
 	}
 	if (!GH(tcache_read)(core, tcache_start, rz_tcache)) {
+		GH(tcache_free)
+		(rz_tcache);
```
Comment (yossizap) [inline librz/core/linux_heap_glibc.c:1139]: Single line?

File: librz/include/rz_heap_glibc.h [hunk 16]
```diff
@@ -282,6 +281,46 @@ typedef struct rz_heap_chunk_list_item {
...
+	MallocState *arena; /* The MallocState for the arena */
+} RzArenaListItem;
+
+typedef struct rz_heap_chunk_simple {
+	ut64 addr; /* Base addr of the chunk*/
+	ut64 prev_size; /* size of prev_chunk*/
+	ut64 size; /* size of chunk */
+	bool NMA; /* flag for NON_MAIN_ARENA */
+	bool PI; /* flag for PREV_INUSE*/
+	bool IM; /* flag for IS_MMAPPED*/
```
Comment (yossizap) [inline librz/include/rz_heap_glibc.h]: Just use the full names here without caps lock, no reason to abbreviate and it's not as readable when you're accessing these fields from Cutter/other files.

File: librz/core/linux_heap_glibc.c (hunk 15 above)
Comment (MalhotraPulak) [reply librz/core/linux_heap_glibc.c:1139]: `clang-format` automatically makes it multi line

**Review by yossizap — APPROVED** [review]

File: librz/core/cmd_linux_heap_glibc.c (hunk 13 above)
Comment (ret2libc) [reply librz/core/cmd_linux_heap_glibc.c:55]: Mmmm I'm not sure.... let it be for now!

**Review by ret2libc — APPROVED** [review]

**Pull request merged**